Add mypy support

Wescoeur opened the pull request.

Nambrok committed:
Add mypy stubs

Co-authored-by: Damien Thenot <damien.thenot@vates.tech>
Co-authored-by: Ronan Abhamon <ronan.abhamon@vates.fr>
Signed-off-by: Damien Thenot <damien.thenot@vates.tech>

Wescoeur committed:
Use importlib instead of imp which is deprecated in python 3.4

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
Replace deprecated calls to distutils.spawn.find_executable

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
Replace deprecated calls to distutils.util.strtobool

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
Fix _locked_load calls compatibility with python 3.10

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
Use static analysis tool (mypy)

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
Fix many invalid escape sequences

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
Fix many invalid escape sequences on regexes

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
Fix override of FileSR.attach

The current attach method of FileSR doesn't correctly override the method of the SR class.
It actually adds a "bind" parameter, which is seen as an error by analyzers like mypy.

The "bind" parameter was added by this commit:
"CA-371791: Fix world readable permissions on EXTSR"

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
Fix override of BaseISCSISR.detach

The current detach method of BaseISCSISR doesn't correctly override the method of the SR class.
It actually adds a "delete" parameter, which is seen as an error by analyzers like mypy.

The "delete" parameter was added by this commit:
"iscsi: Delete LUN on detach of RawISCSI"

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
Modernize Lock class using `staticmethod` decorator

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
Modernize GC using `staticmethod` decorator

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
Modernize RefCounter using `staticmethod` decorator

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
Fix override of VDI.delete in many subclasses

It triggers warns in analyzers like mypy.

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
Define and `details` attr on `Failure` mock

Required by mypy. Attr is referenced in ISOSR.py.

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
Fix util.SRtoXML calls in many drivers

Without this change, mypy triggers an error `var-annoted`:
`Need type annotation for "sr_dict"`

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
Use a specific var for NFS options in ISOSR.attach

Prevent mypy errors when a variable type is changed dynamically from list to string.

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
Fix override of `VDI._do_snapshot`

`cloneOp` must be present. Parameters like `snapType` must must be written in the same way
between the parent class and the child class. Otherwise a linter like mypy may return an error.

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
Prevent mypy errors when a variable type is changed in `BaseISCSISR`

Log without this change on `chappasword` and `incoming_chappassword`:
```
error: Incompatible types in assignment (expression has type "bytes", variable has type "str")  [assignment]
```

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
Prevent wrong mypy error regarding `_linstor` member not set

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
Replace `Dict` variable with `info` in `LVHDSR`

Prevent reuse of the `Dict` symbol from the `typing` module.

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
Prevent bad mypy error in TestMultiLUNISCSISR using formatted-string

Avoid:
```
error: Incompatible types in string interpolation (expression has type "object", placeholder has type "int | float | SupportsInt")  [str-format]
```

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
Simplify FakeSMBSR implementation (remove member vars in class)

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
Use `for session` instead of `for e`

Avoid mypy error:
```
error: Assignment to variable "e" outside except: block  [misc]
```

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
Count correctly IQN sessions during ISCSISR attach

Before this change, IQNs were concatenated into a single string when `multiSession` was used.

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
Fix override of VDI.load in LVHDVDI cleanup.py

It triggers warns in analyzers like mypy.

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

gthvn1 reviewed (approved):
Pretty cool :+1: 
First I was afraid... because I thought that also parameters will be annotated.

Wescoeur commented:
>  First I was afraid... because I thought that also parameters will be annotated.

Only return type is required for a good detection of wrong functions marked as `override`. But I think it's a good idea to use parameter annotation for new modules.

stormi commented:
Is this going to be submitted upstream? Looks like a large diff on many files, so huge potential for rebase conflicts.

Wescoeur commented:
> Is this going to be submitted upstream? Looks like a large diff on many files, so huge potential for rebase conflicts.

At least the first 16 commits. I'm waiting for feedback from our side and to have passed the tests before making an upstream PR.

Nambrok commented on drivers/cleanup.py:
```diff
@@ -1200,27 +1206,31 @@ class LVHDVDI(VDI):
+    @override
+    def load(self, info=None) -> None:
+        # `info` is always set. `None` default value is only here to match parent method.
+        assert info
```
A message for the assert would be good in case it ever happen.

Nambrok commented on tests/test_ISCSISR.py:
```diff
@@ -129,7 +136,7 @@ def setUp(self):
         }
         self.node_records = [(
-            "%s:%d" % (self.node2['ip'], self.node2['port']),
+            f"{self.node2['ip']}:{self.node2['port']}",
```
f-string are only available in python3, would a use of format() make more sense here?

Wescoeur replied:
This PR is only for XCP-ng 8.3. So no issue here. :)

Wescoeur committed:
Makefile fix: don't execute precheck during installation

Due to mypy modifications, we can't build the sm RPM in Koji
without a recent pylint version. So the precheck target is only
executed in a github workflow now.

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
Use `override` everywhere!

Also use return type annotations on these methods.

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

gthvn1 commented on drivers/blktap2.py:
```diff
@@ -2244,7 +2283,7 @@ class XenbusDevice(BusDevice):
 
     XBT_NIL = ""
```
I don't understand why for example XENBUS_DEVTYPE is annotated but XBT_NIL is not?

Wescoeur replied:
The original def is:
```python
XENBUS_DEVTYPE = None
```

Without the XENBUS_DEVTYPE change we have:
```
> mypy .
drivers/blktap2.py:2427: error: Incompatible types in assignment (expression has type "str", base class "XenbusDevice" defined the type as "None")  [assignment]
Found 1 error in 1 file (checked 127 source files)
```

This issue is fixed using:
```python
XENBUS_DEVTYPE: ClassVar[str] = ""
```

In practice we can directly use:
```python
XENBUS_DEVTYPE = ""
```

But I don't know if's better to explicit the type here. Maybe we can rely on type inference?

gthvn1 replied:
If we start to annotate the class variable we need to do it for all of them IMHO. But if it is too much modification we can rely on type inference for now and we will see. The point is that currently we have some of them that are annotated and others that are not so it is a little bit strange. But also as now I understand the point I will not refuse the PR for that :).

gthvn1 reviewed (approved):
LGTM.

gthvn1 commented:
:tada:
